src/kernel.cpp

```diff
@@ -84,7 +87,14 @@ cl_int cvk_kernel::set_arg(cl_uint index, size_t size, const void* value) {
 
     auto const& arg = m_args[index];
 
-    return m_argument_values->set_arg(arg, size, value);
+    cl_int ret = m_argument_values->set_arg(arg, size, value);
+
+    if (ret == CL_SUCCESS) {
+        // Mark argument as set
+        m_args_set[index] = true;
+    }
```

Comment:
We probably want to set the entry to false in the case of a failed argument setting operation to handle the following scenario:
- all arguments set and kernel enqueued successfully
- one argument changed but failed to be set
- subsequent kernel enqueues should fail but they would succeed with the current code.

Probably worth extending the test with that case as well.

Reply:
This is an interesting scenario. I can't see anywhere in the spec that explicitly describes what should happen here. I've just written the test case (passing `0` to the `arg_size` argument of `clSetKernelArg()` and then trying to enqueue the kernel again), but so far all of the native OpenCL implementations I've tried (Mali, Adreno, NVIDIA) still allow the kernel to be enqueued, despite the fact that they all return the correct error from `clSetKernelArg`.

If you think this is the wrong behavior, maybe this needs a spec change before we land this in clvk?

Reply:
Agree we should clarify the spec. I've created https://github.com/KhronosGroup/OpenCL-Docs/issues/548. Would you prefer to merge this change as is and follow up later or wait for the spec issue to be resolved? (I'm fine either way.)

Reply:
There is no urgency to land this PR from my side, happy to wait for resolution on the spec issue.

Reply:
@jrprice We're yet to update the spec but I think the agreement we've reached with the WG means this PR could now get in with a small tweak.

Reply:
@kpet I added a test for enqueuing after a failed call to `clSetKernelArg`, which I believe we concluded will be guaranteed to succeed for the type of failure I used. Is this the tweak you meant?

Reply:
@jrprice Apologies I missed your reply. I _think_ I meant that we should mark the argument as unset when argument setting failed with a code indicating that state may have been modified. `cvk_kernel::set_arg` will have modified state when returning `CL_OUT_OF_RESOURCES` for example. However, looking at this a bit more there are more fundamental issues with error handling and calling `clSetKernelArg` on a kernel for which a previous `clSetKernelArg` call failed with `CL_OUT_OF_RESOURCES` would result in a crash. Since the issue pre-dates this PR and this PR is a net improvement, I'll merge it as is (it's been a while!). Thanks!